algorithms/kernel/dbscan/oneapi/cl_kernels/dbscan_cl_kernels.cl

```diff
@@ -0,0 +1,182 @@
+                algorithmFPType val = fabs(data[id * dim + i] - data[j * dim + i]);
+                algorithmFPType dm  = 1.0;
+                for (int j = 0; j < power; j++)
+                {
+                    dm *= val;
+                }
+                sum += dm;
+            }
+            algorithmFPType ret = sub_group_reduce_add(sum);
+            if (local_id == 0)
+            {
+                cur_dist[j] = ret;
+            }
+        }
+    }
+
+    __kernel void count_neighbors(__global const int * assignments, __global const algorithmFPType * data, int row_id, int chunk_offset,
+                                  int chunk_size, int num_points, algorithmFPType eps, __global const int * queue, __global int * counters,
+                                  __global int * undefCounters) {
+        const int index           = get_global_id(0) * get_num_sub_groups() + get_sub_group_id();
+        const int offset          = index * chunk_size;
+        const int number          = num_points - offset < chunk_size ? num_points - offset : chunk_size;
+        row_id                    = chunk_offset < 0 ? row_id : queue[row_id];
+        const int distance_offset = offset + (chunk_offset < 0 ? 0 : chunk_offset);
+        if (index >= num_points) return;
+        const int size                         = get_sub_group_size();
+        const int id                           = get_sub_group_local_id();
+        __global const algorithmFPType * input = &data[distance_offset];
+        __global const int * assigned          = &assignments[offset];
+        int count                              = 0;
+        int new_count                          = 0;
+        for (int i = id; i < number; i += size)
+        {
+            int nbr_flag     = input[i] <= eps ? 1 : 0;
+            int new_nbr_flag = nbr_flag > 0 && assigned[i] == _UNDEFINED_ && (i + offset != row_id) ? 1 : 0;
+            count += nbr_flag;
+            new_count += new_nbr_flag;
+        }
+        int ret_count     = sub_group_reduce_add(count);
+        int ret_new_count = sub_group_reduce_add(new_count);
+        if (id == 0)
+        {
+            counters[index]      = ret_count;
+            undefCounters[index] = ret_new_count;
+        }
+    }
+
+    __kernel void set_buffer_value(__global int * buffer, int index, int value) {
+        const int global_id = get_global_id(0);
+        const int local_id  = get_local_id(1);
+        if (local_id == 0 & global_id == 0) buffer[index] = value;
+    } __kernel void set_buffer_value_by_queue_index(__global int * queue, __global int * buffer, int pos, int value) {
+        const int global_id = get_global_id(0);
+        const int local_id  = get_local_id(1);
+        if (local_id == 0 & global_id == 0) buffer[queue[pos]] = value;
+    }
+
+    __kernel void count_offsets(__global const int * counters, __global int * offsets, int num_offsets) {
+        if (get_global_id(0) > 0 || get_sub_group_id() > 0) return;
```

Comment:
`get_global_id(0) > 0` probably indicates that you don't need to have 2d range for this kernel